lens/s3repo/repo.go

```diff
@@ -0,0 +1,217 @@
+package s3repo
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-multistore"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/vm"
+	"github.com/filecoin-project/lotus/extern/sector-storage/ffiwrapper"
+	"github.com/filecoin-project/lotus/journal"
+	"github.com/filecoin-project/lotus/lib/bufbstore"
+	"github.com/filecoin-project/lotus/lib/ulimit"
+	marketevents "github.com/filecoin-project/lotus/markets/loggers"
+	"github.com/filecoin-project/lotus/node/impl"
+	"github.com/filecoin-project/lotus/node/impl/full"
+	"github.com/filecoin-project/lotus/node/repo"
+	"github.com/filecoin-project/sentinel-visor/lens"
+	"github.com/filecoin-project/specs-actors/actors/runtime/proof"
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	peer "github.com/libp2p/go-libp2p-peer"
+	"github.com/urfave/cli/v2"
+)
+
+type APIOpener struct {
+	// shared instance of the repo since the opener holds an exclusive lock on it
+	rapi *S3API
+}
+
+func NewAPIOpener(c *cli.Context) (*APIOpener, lens.APICloser, error) {
+	rapi := S3API{}
+
+	if _, _, err := ulimit.ManageFdLimit(); err != nil {
+		return nil, nil, fmt.Errorf("setting file descriptor limit: %s", err)
+	}
+
+	bs, err := NewBlockStore(c.String("repo"))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	r := repo.NewMemory(nil)
+
+	lr, err := r.Lock(repo.FullNode)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	mds, err := lr.Datastore("/metadata")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cs := store.NewChainStore(bs, bs, mds, vm.Syscalls(&fakeVerifier{}), journal.NilJournal())
+
+	headKey, err := bs.(*S3Blockstore).getMasterTsKey(c.Context, safetyLookBack)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	headTs, err := cs.LoadTipSet(*headKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load our own chainhead: %w", err)
+	}
+	if err := cs.SetHead(headTs); err != nil {
+		return nil, nil, fmt.Errorf("failed to set our own chainhead: %w", err)
+	}
+
+	sm := stmgr.NewStateManager(cs)
+
+	rapi.FullNodeAPI.ChainAPI.Chain = cs
+	rapi.FullNodeAPI.ChainAPI.ChainModuleAPI = &full.ChainModule{Chain: cs}
+	rapi.FullNodeAPI.StateAPI.Chain = cs
+	rapi.FullNodeAPI.StateAPI.StateManager = sm
+	rapi.FullNodeAPI.StateAPI.StateModuleAPI = &full.StateModule{Chain: cs, StateManager: sm}
+
+	sf := func() {
+		lr.Close()
+	}
+
+	rapi.Context = c.Context
+	rapi.cacheSize = c.Int("lens-cache-hint")
+	return &APIOpener{rapi: &rapi}, sf, nil
+}
+
+func (o *APIOpener) Open(ctx context.Context) (lens.API, lens.APICloser, error) {
+	return o.rapi, lens.APICloser(func() {}), nil
+}
+
+type S3API struct {
```

Comment:
Will this support ChainGetTipSetByHeight?

Reply:
yep.
The current metadata setup either will happen via the lotus metadata in-memory store which will keep an index to answer this.

Here's a janky re-implementation of that which may be the path of least-resistance here too:
https://github.com/filecoin-project/statediff/blob/master/lib/lazysource.go#L63